### ISSUE TITLE
Different behavior for comparison of integer comparison

### ISSUE THREAD

**Issue by StefanoD** [llvm/llvm-project#138566]
clang-version: 20.1.0
compiler flags: -Wall -O1 -Wextra -Wshadow -Weffc++ -Wno-long-long -pedantic -pedantic-errors -std=c++14

When two variables are multiplied, which leads to integer promotion and the result is not stored into a variable, following example [compiles without problems](https://godbolt.org/z/69v43z9nx):

```cpp
#include <cstdint>

bool square_less_than_method_A(uint16_t a, uint64_t b) {
    // a * a does integer promotion
    return (a * a < b);
}

int main(int, char**) {
    return 0;
}
```

Assigning the product to a variable, results in a [compile error](https://godbolt.org/z/sMso378va):

```cpp
#include <cstdint>

bool square_less_than_method_B(uint16_t a, uint64_t b) {
    const auto a_squared = a * a;
    return a_squared < b;
}

int main(int, char**) {
    return 0;
}
```

```
<source>:5:22: warning: comparison of integers of different signs: 'const int' and 'uint64_t' (aka 'unsigned long') [-Wsign-compare]
    5 |     return a_squared < b;
      |            ~~~~~~~~~ ^ ~
```

The second example should also compile as `a*a` is always positive. So, I get here a false-positive warning.

**Comment by AaronBallman** [comment]
This is expected behavior: the analysis is not dataflow sensitive, so for the second case, we only know there's some `const int` being compared to some `unsigned long long` which is why you get the diagnostic. And we cannot (in general) know whether the comparison will always not be problematic due to sign differences just because the variable is `const` because 1) we don't always know the value of a const object (consider things like `void func(const int *ip)` and working with `*ip`), 2) C and C++ have complex rules that actually allow the value to change if the initial definition of the object was non-`const`. So while a human can discern this is a false positive, it requires a more complex static analysis than the compiler frontend implements.

Closing the issue as WONTFIX. While we aim to avoid false positives as much as we can with frontend diagnostics, doing a flow-sensitive analysis would have unacceptable compile time overhead. This diagnostic is disabled by default exactly because of the false positive rates; you're getting the diagnostic from `-Wextra` which has slightly more tolerance for false positives.

**Comment by chiphogg** [comment]
@AaronBallman, what if we flip it around and ask why the _first_ case does _not_ result in a compile time error?  After all, `a * a` is `int`, and we are comparing a (signed) integer to an unsigned.  We know the _value_ of `a * a` will be non-negative, because `a` is unsigned.  However, very high values for `a` can cause `a * a` to exceed the max for `int`.  If the conversion to `int` happens _before_ the multiplication (as I think it does), then this would be undefined behavior.

Is the lack of compiler error because the only possible outcomes are (a) a correct value, which is known to be unsigned; and, (b) undefined behavior --- and the compiler is free to optimize as if undefined behavior will not occur?

-----

For more context: the reason we are asking about this in the first place is because the `a * a < b` form _did_ [cause a compiler error](https://github.com/aurora-opensource/au/issues/395) on some specific toolchain.  When I dug in, I realized that, technically, yes, it is comparing a signed integer to an unsigned.  I tried to add a CI configuration that would reproduce it, but was unsuccessful in all my attempts.  We ended up [simply fixing the issue](https://github.com/aurora-opensource/au/pull/398) without adding any coverage.  The point of these compiler bugs that @StefanoD filed was to explore whether it's possible to get coverage, or whether (as now seems likely) this was just a weird and hard-to-hit edge case that we won't be able to get any reasonable CI coverage for.

**Comment by matthew-reynolds** [comment]
Gentle bump - @zygoloid @AaronBallman Do you know why it was changed to approximate range analysis in https://reviews.llvm.org/D85778 / cff6dda604cb0548bef5e5951dd1e74536110588? Is it because uint16_t * uint16_t becoming signed is undefined behaviour, so rather than warning about the signed comparison, the compiler is counting on the program to ensure that the uint16_t * uint16_t does not overflow and therefore it assumes it will always be positive?

FWIW GCC also exhibits very similar behaviour here. They state that the lack of warning is because they're choosing to lean on the fact that the signed int overflowing due to uint16_t*uint16_t is UB. See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=120121#c1.

---

This is getting a bit off track, but it would be great if the compiler did issue a specific warning here rather than (sometimes) issuing Wsign-compare. With Wsign-compare, users may be tempted to just do a `static_cast<uint16_t>(a * b)`, which does not address the overflow UB. Some more specific warning about the risk of UB when the multiplication operands are unsigned integer types with width of exactly one half of `unsigned int` could prompt users to properly address the issue, with good ol' `1U * a * b` or by assigning/casting the operands to a type with width >= `unsigned int` prior to the multiplication. Though I feel like this would be a pretty noisy warning for a lot of codebases.

**Comment by shafik** [comment]
This behavior goes back to clang-3: https://godbolt.org/z/chbbGP6bn

So this is not a recent change. 

I think this behavior is rather unfortunate but maybe not easy to fix. `uint16_t` multiplication is one of those really odd edge cases: https://shafik.github.io/c++/2021/12/30/usual_arithmetic_confusions.html

Working w/ mixed integer types is tricky, I think explicit conversions and strict assertions of preconditions is the way to go here.

**Comment by matthew-reynolds** [comment]
> This behavior goes back to clang-3: https://godbolt.org/z/chbbGP6bn
>
> So this is not a recent change.

Ah, I see. Seems that the approximate range analysis may have been a red herring then. I also recompiled clang forcing `Approximate` to `false` in `TryGetExprRange`, and saw the same behaviour (No warning for `a * a < b`), so definitely seems like this is not the root of the issue. So back to @chiphogg's question - Do we know offhand why the `a * a < b` does not warn? 

> I think this behavior is rather unfortunate but maybe not easy to fix. `uint16_t` multiplication is one of those really odd edge cases: https://shafik.github.io/c++/2021/12/30/usual_arithmetic_confusions.html
>
> Working w/ mixed integer types is tricky, I think explicit conversions and strict assertions of preconditions is the way to go here.

Definitely, it's just a bummer that the compiler doesn't warn on it.
